### PR TITLE
Add a utility to run go benchmarks and output JUnit XML.

### PR DESCRIPTION
**Pull request by cjwagner**
This also includes
- a `dummybenchmarks` package for testing/demoing
- a canary job for validating JUnit compatibility with spyglass and Testgrid.

I looked into using both `go test`'s `-json` flag and https://github.com/jstemmer/go-junit-report, but neither were suitable.
<details><summary>my notes</summary>

go-junit-report:
- doesn’t expose skipped benchmarks
- doesn’t expose op count, alloc, or throughput data.
    - And no way to easily inject additional properties per testcase.
- doesn’t include logged output per benchmark testcase.
- if any benchmark fails all output (including from passing benchmarks) is put in a single “Failure” testcase.
    - Ideally each failed benchmark would be individually represented with a testcase and only include its own failure text.

`go test -json` output:
- Doesn’t identify benchmarks with an `action: bench` entry unless it logs something on its own. Otherwise have to parse the raw output.
- Doesn’t include any timing data or other metrics
- Basically useless.

</details>

Sample `go test -bench` output and corresponding JUnit XML:

<details><summary>samples</summary>

```
goos: darwin
goarch: amd64
pkg: k8s.io/test-infra/experiment/dummybenchmarks
BenchmarkSimple-4           	10000000	       138 ns/op
Benchmark-4                 	10000000	       129 ns/op
BenchmarkReportAllocs-4     	10000000	       134 ns/op	     112 B/op	       2 allocs/op
BenchmarkSetBytes-4         	10000000	       132 ns/op	 151.01 MB/s
BenchmarkAllocsAndBytes-4   	10000000	       149 ns/op	 133.83 MB/s	     112 B/op	       2 allocs/op
BenchmarkParallel-4         	20000000	        76.2 ns/op
BenchmarkLog-4              	10000000	       139 ns/op
--- BENCH: BenchmarkLog-4
    dummy_test.go:67: About to DoTheThing() x1.
    dummy_test.go:67: About to DoTheThing() x100.
    dummy_test.go:67: About to DoTheThing() x10000.
    dummy_test.go:67: About to DoTheThing() x1000000.
    dummy_test.go:67: About to DoTheThing() x10000000.
--- SKIP: BenchmarkSkip
    dummy_test.go:74: This Benchmark is skipped.
--- SKIP: BenchmarkSkipNow
--- FAIL: BenchmarkError
    dummy_test.go:82: Early Benchmark error.
    dummy_test.go:67: About to DoTheThing() x1.
--- FAIL: BenchmarkFatal
    dummy_test.go:87: This Benchmark failed.
--- FAIL: BenchmarkFailNow
BenchmarkNestedShallow/simple-4         	10000000	       134 ns/op
BenchmarkNestedShallow/parallel-4       	20000000	        77.4 ns/op
BenchmarkNestedDeep/1_simple-4          	10000000	       129 ns/op
BenchmarkNestedDeep/1_parallel-4        	20000000	        75.4 ns/op
BenchmarkNestedDeep/3A_simple-4         	10000000	       130 ns/op
BenchmarkNestedDeep/3A-4                	       1	1437896682 ns/op
BenchmarkNestedDeep/3B_parallel-4       	20000000	        76.1 ns/op
BenchmarkNestedDeep/3B-4                	       1	1598391068 ns/op
BenchmarkNestedDeep/2-4                 	       1	3036768685 ns/op
BenchmarkNestedDeep/1-4                 	       1	6048035011 ns/op
FAIL
exit status 1
FAIL	k8s.io/test-infra/experiment/dummybenchmarks	19.862s
goos: darwin
goarch: amd64
pkg: k8s.io/test-infra/experiment/dummybenchmarks/subpkg
BenchmarkSubPkg-4   	10000000	       129 ns/op
PASS
ok  	k8s.io/test-infra/experiment/dummybenchmarks/subpkg	1.477s
```

```xml
<testsuites>
    <testsuite name="k8s.io/test-infra/experiment/dummybenchmarks" tests="20" failures="3" time="19.974000">
        <testcase class_name="dummybenchmarks" name="BenchmarkSimple-4" time="1.400000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="140"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="Benchmark-4" time="1.580000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="158"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkReportAllocs-4" time="1.400000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="140"></property>
                <property name="B/op" value="112"></property>
                <property name="allocs/op" value="2"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkSetBytes-4" time="1.310000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="131"></property>
                <property name="MB/s" value="152.04"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkAllocsAndBytes-4" time="1.300000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="130"></property>
                <property name="MB/s" value="152.69"></property>
                <property name="B/op" value="112"></property>
                <property name="allocs/op" value="2"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkParallel-4" time="1.524000">
            <properties>
                <property name="op count" value="20000000"></property>
                <property name="avg op duration (ns/op)" value="76.2"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkLog-4" time="1.290000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="129"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkSkip" time="0">
            <skipped>true</skipped>
            <properties></properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkSkipNow" time="0">
            <skipped>true</skipped>
            <properties></properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkError" time="0">
            <failure>    dummy_test.go:82: Early Benchmark error.&#xA;    dummy_test.go:67: About to DoTheThing() x1.&#xA;</failure>
            <properties>
                <property name="categorized_fail" value="    dummy_test.go:82: Early Benchmark error.&#xA;    dummy_test.go:67: About to DoTheThing() x1.&#xA;"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkFatal" time="0">
            <failure>    dummy_test.go:87: This Benchmark failed.&#xA;</failure>
            <properties>
                <property name="categorized_fail" value="    dummy_test.go:87: This Benchmark failed.&#xA;"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkFailNow" time="0">
            <failure></failure>
            <properties></properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedShallow/simple-4" time="1.290000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="129"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedShallow/parallel-4" time="1.542000">
            <properties>
                <property name="op count" value="20000000"></property>
                <property name="avg op duration (ns/op)" value="77.1"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/1_simple-4" time="1.300000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="130"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/1_parallel-4" time="1.526000">
            <properties>
                <property name="op count" value="20000000"></property>
                <property name="avg op duration (ns/op)" value="76.3"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/3A_simple-4" time="1.320000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="132"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/level_3A-4" time="1.471533">
            <properties>
                <property name="op count" value="1"></property>
                <property name="avg op duration (ns/op)" value="1471532694"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/3B_parallel-4" time="1.524000">
            <properties>
                <property name="op count" value="20000000"></property>
                <property name="avg op duration (ns/op)" value="76.2"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/level_3B-4" time="1.602681">
            <properties>
                <property name="op count" value="1"></property>
                <property name="avg op duration (ns/op)" value="1602680853"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/level_2-4" time="3.074794">
            <properties>
                <property name="op count" value="1"></property>
                <property name="avg op duration (ns/op)" value="3074794260"></property>
            </properties>
        </testcase>
        <testcase class_name="dummybenchmarks" name="BenchmarkNestedDeep/level_1-4" time="6.111073">
            <properties>
                <property name="op count" value="1"></property>
                <property name="avg op duration (ns/op)" value="6111073302"></property>
            </properties>
        </testcase>
    </testsuite>
    <testsuite name="k8s.io/test-infra/experiment/dummybenchmarks/subpkg" tests="1" failures="0" time="1.482000">
        <testcase class_name="subpkg" name="BenchmarkSubPkg-4" time="1.300000">
            <properties>
                <property name="op count" value="10000000"></property>
                <property name="avg op duration (ns/op)" value="130"></property>
            </properties>
        </testcase>
    </testsuite>
</testsuites>
```

</details>

/assign @fejta @Katharine
/cc @stevekuznetsov

### CODE REVIEW

**Comment by cjwagner** [comment]
I put the JUnit benchmark results for `dummybenchmarks` into the `artifacts` directory of a run of ci-test-infra-canary-echo-test to confirm that the JUnit parses correctly in Testgrid and spyglass. See the 7:28pm (PST) run:

https://testgrid.k8s.io/sig-testing-canaries#echo-test&graph-metrics=MB%2Fs&graph-metrics=avg%20op%20duration%20(ns%2Fop)&graph-metrics=alloced%20B%2Fop&width=90&graph-metrics=allocs%2Fop&graph-metrics=test-duration-minutes (The `BenchmarkAllocsAndBytes-4` test demonstrates all the supported metric types.)

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-test-infra-canary-echo-test/1121964406209515522

File: config/jobs/kubernetes/test-infra/test-infra-canaries.yaml [hunk 1]
```diff
@@ -33,3 +33,22 @@ periodics:
...
+  - name: ci-test-infra-benchmark-demo
+    interval: 20m
+    decorate: true
+    extra_refs:
+    - org: cjwagner
+      repo: test-infra
+      base_ref: microbe-nchmarks
+    spec:
+      containers:
+      - image: golang:latest
```
Comment (fejta) [inline config/jobs/kubernetes/test-infra/test-infra-canaries.yaml]: ❤️

File: pkg/benchmarkjunit/parse.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	PropertyList []Property `xml:"property"`
+}
+
+// Failure defines the xml element that stores failure text.
+type Failure struct {
+	Text string `xml:",chardata"`
+}
+
+// TestCase defines the xml element that stores all information associated with one benchmark.
+type TestCase struct {
```
Comment (fejta) [inline pkg/benchmarkjunit/parse.go]: We don't want to use/extend https://github.com/kubernetes/test-infra/blob/master/testgrid/metadata/junit/junit.go?

File: pkg/benchmarkjunit/parse.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	reBenchMetrics = regexp.MustCompile(`^(Benchmark\S*)\s+(\d+)\s+([\d\.]+) ns/op(?:\s+([\d\.]+) MB/s)?(?:\s+([\d\.]+) B/op)?(?:\s+([\d\.]+) allocs/op)?\s*$`)
+	// reActionLine identifies lines that start with "--- " and denote the start of log output and/or a skipped or failed Benchmark.
+	// Matches lines like:
+	// "--- BENCH: BenchmarkLog-4"
+	// "--- SKIP: BenchmarkSkip"
+	// "--- FAIL: BenchmarkFatal"
+	reActionLine = regexp.MustCompile(`^--- (BENCH|SKIP|FAIL):\s+(\S+)\s*$`)
+)
+
+// Property defines the xml element that stores additional metrics about each benchmark.
```
Comment (stevekuznetsov) [inline pkg/benchmarkjunit/parse.go]: Can't we import these types from somewhere?

File: pkg/benchmarkjunit/parse.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+		props = append(props, Property{Name: "alloced B/op", Value: match[5]})
+	}
+	if len(match[6]) > 0 {
+		props = append(props, Property{Name: "allocs/op", Value: match[6]})
+	}
+
+	return &Properties{PropertyList: props}, benchmarkDuration, nil
+}
+
+func parse(raw []byte) (*TestSuites, error) {
```
Comment (stevekuznetsov) [inline pkg/benchmarkjunit/parse.go]: Since you have the testdata dir, we could write an integration test if you wanted

**Comment by cjwagner** [comment]
I added a README, and an integration test.
I also switched to using the existing types defined in `testgrid/metadata/junit`.

**Comment by cjwagner** [comment]
Finally made bazel happy! PTAL.
/hold cancel

**Comment by fejta** [comment]
What was the secret?

File: testgrid/metadata/junit/junit.go [hunk 5]
```diff
@@ -44,15 +44,47 @@ type Suite struct {
...
+	Error      *string     `xml:"system-err,omitempty"`
+	Skipped    *string     `xml:"skipped,omitempty"`
+	Properties *Properties `xml:"properties,omitempty"`
+}
+
+func (r *Result) SetProperty(name, value string) {
+	if r.Properties == nil {
+		r.Properties = &Properties{}
+	}
+	for i, existing := range r.Properties.PropertyList {
```
Comment (michelle192837) [inline testgrid/metadata/junit/junit.go]: Short comment noting this replaces an existing property?

**Comment by cjwagner** [comment]
> What was the secret?

To get past `compile: version "go1.12.1" does not match go tool version "go1.12.2"` I had to explicitly set GOROOT to the `go_sdk` dir from the integration test. This envvar was undefined in CI and as a result go was looking for tools (like compile) in the `/usr/local/go` dir even though the go binary being used was from bazel's `go_sdk` dir.
To actually include all the tools, standard library, and header files I had to add `"@go_sdk//:files"` as a data dependency of the `go_default_test` bazel rule.

File: testgrid/metadata/junit/junit.go (hunk 5 above)
Comment (cjwagner) [reply testgrid/metadata/junit/junit.go]: Done.

**Comment by cjwagner** [comment]
https://testgrid.k8s.io/sig-testing-canaries#benchmark-demo&width=20&graph-metrics=avg%20op%20duration%20(ns%2Fop)&graph-metrics=test-duration-minutes&graph-metrics=op%20count

**Pull request merged**